### DISCUSSION TITLE
Kyuubi ODBC Example

### DISCUSSION THREAD

**Discussion by JeffSBailey** [apache/kyuubi#5355] [Q&A]
I have successfully been able to connect my application to Kyuubi using a myssql connection (marked as experimental in the Helm chart).  So far, this does not allow table discovery from my application.  I was going to try to connect via ODBC but I cannot find any examples of how to setup an ODBC DSN to connect.  Does anyone have any examples they could share?

**Comment by pan3793** [comment]
Sorry, there are no ODBC docs yet, would appreciate it if you could send PR to supply one.

It's recommended to use [Databricks Spark ODBC driver](https://www.databricks.com/spark/odbc-drivers-download) to connect Kyuubi.

Technical details: both JDBC/ODBC API provided by Kyuubi are based on Thrift protocol, it's fully compatible with Apache Hive Thrift protocol(Impala, and Spark Thrift Server did it too), so ideally, you can use any Hive client to connect Kyuubi. However, some ODBC clients may reject the connection if the GetInfo result returned by the server is not recognized, e.g. server version.

**Reply by AlexSpecflow** [reply]
Hi.

raising this old question. Is there any progress on getting support of connectivity to Kyuubi through an ODBC driver?
My case is to query Kyuubi from a .net core application (.net 8).
How can i do that? Is there any cost-free or commercial ODBC driver that allows to connect to Kyuubi ?

https://kyuubi.readthedocs.io/en/v1.11.0/index.html
In the documentation is mentioned this:
_A HiveServer2-compatible interface that allows end users to use a thrift client(cross-language support, both tcp and http), a [Java Database Connectivity(JDBC)](https://kyuubi.readthedocs.io/en/v1.11.0/client/jdbc/index.html) interface over thrift, or an [Open Database Connectivity (ODBC)](https://kyuubi.readthedocs.io/en/v1.11.0/client/odbc/index.html) interface over a JDBC-to-ODBC bridge to communicate with Kyuubi._

Given this - i tried "Cloudera ODBC Driver for Apache Hive" from my .net 8 app (which BTW successfully retrieves data from Hive), but got:
`System.Data.Odbc.OdbcException HResult=0x80131937 Message=ERROR [HY000] [Cloudera][DriverSupport] (1110) Unexpected response received from server. Please ensure the server host and port specified for the connection are correct and confirm if SSL should be enabled for the connection. ERROR [HY000] [Cloudera][DriverSupport] (1110) Unexpected response received from server. Please ensure the server host and port specified for the connection are correct and confirm if SSL should be enabled for the connection.[Cloudera][DriverSupport] (1110) Unexpected response received from server. Please ensure the server host and port specified for the connection are correct and confirm if SSL should be enabled for the connection. Source= StackTrace: at System.Data.Odbc.OdbcConnection.HandleError(OdbcHandle hrHandle, SQLRETURN retcode) at System.Data.Odbc.OdbcConnectionHandle..ctor(OdbcConnection connection, OdbcConnectionString constr, OdbcEnvironmentHandle environmentHandle) at System.Data.Odbc.OdbcConnectionOpen..ctor(OdbcConnection outerConnection, OdbcConnectionString connectionOptions)`

However, i am able to retrieve data successfully using "beeline" tool or Dbeaver through the JDBC interface (connection parameters like host, port, etc - are the same as in my unsuccessful attempt to get data from the .net app).

PS: i also tried "Databricks ODBC Driver", but the result is the same.

**Reply by pan3793** [reply]
We got several reports that users successfully use [Databricks Spark ODBC driver](https://www.databricks.com/spark/odbc-drivers-download) to connect Kyuubi. But given it's not open-source software, we don't know much of the underlying implementation details, thus have no ideas on error you reported.

**Reply by AlexSpecflow** [reply]
is there any example of how connectionString must be built and C# code snippet demonstrating that? + ideally a docker-compose file that would set up Kyuubi instance along with needed dependencies.
